### PR TITLE
Further tests for Pools and some sizing fixes

### PR DESCRIPTION
**Pull request by adamretter**
Closes https://github.com/eXist-db/exist/pull/4047
Closes https://github.com/eXist-db/exist/issues/4045

### CODE REVIEW

**Comment by dizzzz** [comment]
from what I read in the code, things go wrong after returning objects all objects to the pool, and then retrieve an additional one.

**Comment by line-o** [comment]
Does this also fix #4045 ?

File: exist-core/src/main/java/org/exist/util/serializer/SerializerPool.java [hunk 1]
```diff
@@ -51,7 +51,7 @@ private static GenericKeyedObjectPoolConfig toConfig(final int maxIdle) {
         config.setBlockWhenExhausted(false);
         config.setLifo(true);
         config.setMaxIdlePerKey(maxIdle);
-        config.setMaxTotalPerKey(-1);       // TODO(AR) is this the best way to allow us to temporarily exceed the size of the pool?
+//        config.setMaxTotalPerKey(-1);       // TODO(AR) is this the best way to allow us to temporarily exceed the size of the pool?
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/util/serializer/SerializerPool.java:54]: Should this line be removed?

**Comment by line-o** [comment]
This looks very good. I have two minor questions.

**Review by dizzzz — APPROVED** [review]

**Review by dizzzz — DISMISSED** [review]
build fails :-(

image

**Comment by adamretter** [comment]
> Does this also fix #4045 ?

@line-o Yes indeed it does.

File: exist-core/src/main/java/org/exist/util/serializer/SerializerPool.java (hunk 1 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/util/serializer/SerializerPool.java]: Yes, I have now corrected that.

**Comment by duncdrum** [comment]
It's unfortunate that the bug which seems to be on `develop` as well, is crashing restore tests on windows. I'd normally say that this actually warrants a `hotfix` to ensure ppl can restore into the latest stable release without errors. 

Other then the problems revealed by CI, the changes look good, and i can confirm that the fix java based restores also for docker.

File: exist-core/src/main/java/org/exist/storage/serializers/XmlSerializerPool.java [hunk 2]
```diff
@@ -26,18 +26,21 @@
...
 
-    private static GenericObjectPoolConfig<Serializer> toConfig(final String brokerId, final int maxIdle) {
+    private static GenericObjectPoolConfig<Serializer> toConfig(@Nullable final String brokerId, final int maxIdle) {
         final GenericObjectPoolConfig<Serializer> config = new GenericObjectPoolConfig<>();
         config.setBlockWhenExhausted(false);
         config.setLifo(true);
         config.setMaxIdle(maxIdle);
         config.setMaxTotal(-1);            // TODO(AR) is this the best way to allow us to temporarily exceed the size of the pool?
-        config.setJmxNameBase("org.exist.management.exist:type=XmlSerializerPool,name=pool." + brokerId);
+        final String poolName = brokerId == null ? "" : "." + brokerId;
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/storage/serializers/XmlSerializerPool.java:42]: Shouldn't this be
```java
final String poolName = brokerId == null ? "" : "pool." + brokerId;
```
?

Comment (adamretter) [reply exist-core/src/main/java/org/exist/storage/serializers/XmlSerializerPool.java]: Good catch! I just pushed a commit to address that.

Comment (line-o) [reply exist-core/src/main/java/org/exist/storage/serializers/XmlSerializerPool.java]: 👍🏼

**Review by line-o — APPROVED** [review]
Tests green; Good to go

**Comment by adamretter** [comment]
@line-o Thanks, would you mind merging please?

File: exist-core/src/main/java/org/exist/storage/BrokerPool.java [hunk 3]
```diff
@@ -1095,11 +1095,25 @@ public PerformanceStats getPerformanceStats() {
...
     public XMLReaderPool getParserPool() {
         return xmlReaderPool;
     }
 
+    /**
+     * Returns a pool in which the database instance's readers are stored.
+     *
+     * @return The pool
+     *
+     * @deprecated Use {@link #getXmlReaderPool()} instead
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/storage/BrokerPool.java:1111]: I wonder if this should not have the deprecated field in the docblock referencing itself. At least it is confusing me.

**Pull request merged**